Post about rtweet (#362)
ropensci/roweb3: content/blog/2022-07-10-rtweet-1-0-0/index.md

---
title: 'Upgrading rtweet'
author:
  - Lluís Revilla Sancho
date: '2022-07-10'
slug: rtweet-1-0-0
categories:
  - blog
tags:
  - rtweet
  - R
package_version: 1.0.0
description: Update from rtweet 0.7.0 to 1.0.0.
twitterImg: blog/2019/06/04/post-template/name-of-image.png

> twitterAlt: Alternative description of the image

Author: I wasn't sure which one to add. I thought that maybe there would be some suggestion. But I agree that leaving this is not helpful, so I'll remove it

However, to preserve the nested nature of the data returned some fields are now nested inside other.
Some columns previously provided by rtweet are now not returned.
At the same time it provides with new columns about each tweet like the `withheld_*` columns.

If you scanned through the columns you might have noticed that there aren't `"user_id"` or `"screen_name"` columns returned.
This data is still returned by the API but it is now made available to the rtweet users via `users_data()`:

```r
> colnames(users_data(tweets))
 [1] "id"                      "id_str"                 
 [3] "name"                    "screen_name"            
 [5] "location"                "description"            
 [7] "url"                     "protected"              
 [9] "followers_count"         "friends_count"          
[11] "listed_count"            "created_at"             
[13] "favourites_count"        "verified"               
[15] "statuses_count"          "profile_image_url_https"
[17] "profile_banner_url"      "default_profile"        
[19] "default_profile_image"   "withheld_in_countries"  
[21] "derived"                 "withheld_scope"         
[23] "entities" 
```

> 

Author: I'll add this example.

> This should help finding the right columns of data but if you don't find what you are looking for it might be nested inside a column. 

Author: Yes, I think it is too much to show:

```r
Rows: 100
Columns: 43
$ created_at                    <dttm> 2022-07-20 10:20:14, 2022-07-19 16:19:0…
$ id                            <dbl> 1.549671e+18, 1.549398e+18, 1.549298e+18…
$ id_str                        <chr> "1549670503215939585", "1549398439439605…
$ full_text                     <chr> "11,500 firefighters cut since 2010 + ex…
$ truncated                     <lgl> FALSE, FALSE, FALSE, FALSE, FALSE, FALSE…
$ display_text_range            <dbl> 208, 220, 273, 140, 84, 140, 103, 101, 6…
$ entities                      <list> [[<data.frame[1 x 2]>], [<data.frame[1 …
$ metadata                      <list> [<data.frame[1 x 2]>], [<data.frame[1 x…
$ source                        <chr> "<a href=\"http://twitter.com/download/a…
$ in_reply_to_status_id         <dbl> NA, NA, NA, NA, NA, NA, 1.549840e+18, NA…
$ in_reply_to_status_id_str     <chr> NA, NA, NA, NA, NA, NA, "154983985355420…
$ in_reply_to_user_id           <dbl> NA, NA, NA, NA, NA, NA, 3.410549e+08, NA…
$ in_reply_to_user_id_str       <chr> NA, NA, NA, NA, NA, NA, "341054873", NA,…
$ in_reply_to_screen_name       <chr> NA, NA, NA, NA, NA, NA, "matt_greener", …
$ geo                           <list> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,…
$ coordinates                   <list> [<data.frame[1 x 3]>], [<data.frame[1 x…
$ place                         <list> [<data.frame[1 x 3]>], [<data.frame[1 x…
$ contributors                  <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ is_quote_status               <lgl> FALSE, FALSE, TRUE, FALSE, FALSE, FALSE,…
$ retweet_count                 <int> 2725, 3727, 3580, 5, 253, 9, 0, 0, 2, 0,…
$ favorite_count                <int> 7235, 6402, 5616, 0, 0, 0, 0, 0, 0, 0, 0…
$ favorited                     <lgl> FALSE, FALSE, FALSE, FALSE, FALSE, FALSE…
$ retweeted                     <lgl> FALSE, FALSE, FALSE, FALSE, FALSE, FALSE…
$ lang                          <chr> "en", "en", "en", "en", "en", "en", "en"…
$ possibly_sensitive            <lgl> NA, FALSE, FALSE, FALSE, FALSE, NA, NA, …
$ quoted_status_id              <dbl> NA, NA, 1.549277e+18, NA, NA, NA, NA, NA…
$ quoted_status_id_str          <chr> NA, NA, "1549277296397352961", NA, NA, N…
$ quoted_status                 <list> [<data.frame[1 x 27]>], [<data.frame[1 …
$ retweeted_status              <list> [<data.frame[1 x 27]>], [<data.frame[1 …
$ text                          <chr> "11,500 firefighters cut since 2010 + ex…
$ favorited_by                  <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ scopes                        <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ display_text_width            <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ quoted_status_permalink       <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ quote_count                   <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ timestamp_ms                  <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ reply_count                   <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ filter_level                  <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ query                         <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ withheld_scope                <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ withheld_copyright            <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ withheld_in_countries         <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ possibly_sensitive_appealable <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
```
I would also like to avoid providing information from tweets.

There are "few" nested columns: entities, metadata, geo, coordinates, place, quoted_status, retweeted_status. Perhaps it is easier and still useful to mention them?

> This should help finding the right columns of data but if you don't find what you are looking for it might be nested inside a column. 

Reviewer: Good point about not including tweet data. Since this is still a good way to explore the data, perhaps use something like: 

Use `glimpse()` to explore the data and see which columns are nested. 

Then you don't have to list the columns or show too much?

> This should help finding the right columns of data but if you don't find what you are looking for it might be nested inside a column. 

Author: Good idea! Thanks also for introducing me to pillar.

more_users <- get_followers("_R_Foundation", cursor = users)
```

They support `max_id` and `since_id` to find earlier and later tweets respectively:

```r
# Retrieve all the tweets made since the previous request
newer <- search_tweets("weather", since_id = tweets)
# Retrieve tweets made before the previous request
older <- search_tweets("weather", max_id = tweets)
```

If you want more tweets than it is allowed by the rate limits of the API, you can use `retryonratelimit` to wait as long as needed:

```r
long <- search_tweets("weather", n = 1000, retryonratelimit = TRUE)
```

This will keep busy your terminal until the 1000 tweets are retrieved.

### Saving data

An unexpected consequence of returning more data as in the API is that it is now harder to save it in a tabular format.
For instance one tweet might have one media, mention two user and have three hasthags.
There isn't a simple way to saved it in a single row uniformly for all tweets or 
it could lead to confusion.

This resulted in deprecating `save_as_csv`, `read_twitter_csv` and related functions because it doesn't work with the new data structure and it won't be possible to load it from a csv. 

> They will be removed in later versions.

Author: Many thanks for the suggestion. I wasn't aware of `write_rds`

> ## **Other breaking changes**

Author: Indeed, I had my doubts when I wrote them.

Accessibility is important for that reason if you tweet via `post_tweet()` and add an image, gif or video you'll need to provide the media alternative text. 
Without `media_alt_text` it will not allow you to post.

`tweet_shot()` has been deprecated as it did no longer work correctly. 
There might be possible to bring it back, but I don't understand the code provided and can't maintain it (see the long discussion about it [on the issue](https://github.com/ropensci/rtweet/issues/458)). 
  

> As emojis, langs and stopwordslangs data provided with rtweet are useful resources for text mining in general - not only in tweets - and they need to be updated to be helpful, they are no longer available on the package.

Author: I think emojis found a new place in a package that previously depended on rtweet for it. But they are not functions, so it cannot be called via `()` But I'll merge your suggestion and modify later.

The functions like `suggested_*()` have been removed as they didn't work since 2019.

## **Easier authentication**

As part of this release there have been a big rewrite of the authentication protocol.

> While it is version compatible it has also some new functions that are important as these functions make it easier for different use cases to work with rtweet and the  API.

Author: It is compatible with previous versions of rtweet, so authentication method in 0.7.0 works in 1.0.0.

It will also automatically move your tokens to the new path.

To check which credentials you have stored use `auth_list()` and load them via `auth_as("my_app")`.
All the rtweet functions will use the latest token loaded with `auth_as` (unless you manually specify one when calling it).
If you are not sure which token you are using you can use `auth_get()` it will return the token in use, list them or ask you to authenticate.

## **Other changes**

This is a list of other changes that aren't too big or are not breaking changes but are worthy enough of a mention:

### Iteration and continuation of requests

Using cursors, pagination or waiting until you can make more queries is now easier.
For example you can continue previous requests via:

```r
users <- get_followers("_R_Foundation")
users

# use `cursor` to find the next "page" of results
more_users <- get_followers("_R_Foundation", cursor = users)
```

### Additions

There is now a function to find a thread of a user.
You can start from any tweet and it will find all the tweets of the thread:
`tweet_threading("1461776330584956929")`.

> The interest on download and keep track of interactions in twitter is big. 

Author: Done, currently it is as: 

> There is a lot of interest in downloading and keeping track of interactions on Twitter.The amount of interest is big enough that Twitter is releasing a new API to provide more information of this nature.

```

They support `max_id` and `since_id` to find earlier and later tweets respectively:

```r
# Retrieve all the tweets made since the previous request
newer <- search_tweets("weather", since_id = tweets)
# Retrieve tweets made before the previous request
older <- search_tweets("weather", max_id = tweets)
```

If you want more tweets than it is allowed by the rate limits of the API, you can use `retryonratelimit` to wait as long as needed:

```r
long <- search_tweets("weather", n = 1000, retryonratelimit = TRUE)
```

This will keep busy your terminal until the 1000 tweets are retrieved.

### Saving data

An unexpected consequence of returning more data as in the API is that it is now harder to save it in a tabular format.
For instance one tweet might have one media, mention two user and have three hasthags.
There isn't a simple way to saved it in a single row uniformly for all tweets or 
it could lead to confusion.

This resulted in deprecating `save_as_csv`, `read_twitter_csv` and related functions because it doesn't work with the new data structure and it won't be possible to load it from a csv. 
They will be removed in later versions.

## **Other breaking changes**

> If you tweet via `post_tweet()` and add an image, gif or video you'll need to provide the media alternative text. 

Author: I used your suggestion removing "to our community", as it is not only important here ;)

- Set up a bot: `rtweet_bot()`

Find more information on the [Authentication with rtweet vignette](https://docs.ropensci.org/rtweet/articles/auth.html).

Previously rtweet saved each token created, if you don't use the default token you'll need to save them manually via `auth_save(token, "my_app")`. 
Bonus, if you name your token as default (`auth_save(token, "default")`) it will be used since loading the library.

The tokens are now not saved on your home directory but via `tools::R_user_dir("rtweet", "config")`. 
If you have previous tokens saved or problems identifying which token is which use `auth_sitrep()`.
This will provides clues to which tokens might be duplicated or misconfigurated but it won't check if they work. 
It will also automatically move your tokens to the new path.

To check which credentials you have stored use `auth_list()` and load them via `auth_as("my_app")`.
All the rtweet functions will use the latest token loaded with `auth_as` (unless you manually specify one when calling it).
If you are not sure which token you are using you can use `auth_get()` it will return the token in use, list them or ask you to authenticate.

> ## **Other changes**

Author: I like more the second one, I hope all changes are interesting :smile: .

> Some other changes aren't too big to be worth their own section or are not breaking changes but are worthy enough of a mention:

Reviewer 2: ```suggestion
This is a list of other changes that aren't too big or are not breaking changes but are worthy enough of a mention:
```

It is big enough that Twitter is releasing a new API to provide more information about this.

## **Future**

Twitter API v2 is being released and soon it will replace API v1. 
rtweet up to now, including this release, uses API v1 so it will need to adapt to the new endpoints and new data returned.

First will be the streaming endpoints in November, so expect more (breaking?) changes around those dates if not earlier.

I would also like to make it easier for users, dependencies and the package itself to handle the outputs.
To this regard I would like to provide some classes to handle the different type of objects it returns.

This would help to avoid some current shortcomings. 
Specifically I would like to provide functions to make it easier to reply to previous tweets, 
provide shortcuts to find out some nested information and subset tweets and the accompanying user information.

## **Conclusions**

While I made many breaking changes I hope this helps both users and maintainers.

Feel free to ask on the [rOpenSci community](https://discuss.ropensci.org/tag/rtweet) if you have questions about the transition or find something amiss. 

> Please let me know, it will help me prioritize which endpoints I are more relevant to the community. 

Reviewer 2: ```suggestion
Please let me know, it will help me prioritize which endpoints are more relevant to the community. 
```
